### DISCUSSION TITLE
Differences between 8266 and ESP32-S3 I2C responsiveness?

### DISCUSSION THREAD

**Discussion by papakpmartin** [arendst/Tasmota#20689] [Q&A] [answered]
I have a little _"garage door opened/closed and car presence via time-of-flight"_ thing that I made. "Mark I" was with an 8266, which worked great until the board died after maybe a year. I rebuilt a "Mark II" with a custom ESP32 board that I have which has a few sensors on two I2C busses, all on the `0` core (along with a relay and a PWM RGB LED.)

On the 8266, the ToF sensor reported out changes very quickly (1-2 seconds). On the 32 it's much more sluggish (30-60 seconds).

On the 8266, it was one core, one I2C bus. On the 32 it's a couple of sensors on one I2C bus (at GPIO 33 and 34 for SDA and SCL respectively) and the time of flight sensor (a VL53L0X) on the second I2C bus (at GPIO 35 and 36 for SDA and SCL respectively)... all on core `0`.

Should I be doing something special for the second I2C bus to be more responsive?

Bonus (maybe related) question: Should I be maybe be setting up the sensing on core `1` and leave `0` alone for network and messaging? (When I hand-code for these boards, I usually do something like that.)

Thoughts appreciated. Thanks! :)

**Comment by barbudor** [comment]
Hi
There should be no differences between ESP8266 and any ESP32 variant (S3 or others)
I would mostly suspect that you didn't copied the configuration identically from the 8266 to the 32

I have an similar setup with VL53L1X , 2 micro-switch inputs to know if the door is opened or closed and a relay to "push" on the door openner button to be able to remotely open/close the door. The VL53L1X is on the ceiling in the middle of the garage giving ~210cm when no car is there and ~40cm when the car is there.

I have a teleperiod of 3600 in order to periodically push the state of the system to MQTT from where Node-RED feeds an influxDB database
Then I have some rules which performs "translation" of distance into publication to a `PRESENCE` topic with `OFF` or `ON` payload

Here is my set of rules regarding that:
```
08:30:23.702 CMD: rule0
Rule1
   on VL53L1X#Distance<%mem1% do backlog0 ruletimer1 %mem2%; ruletimer2 0 break 
   on VL53L1X#Distance>%mem1% do backlog0 ruletimer2 %mem2%; ruletimer1 0 endon
Rule1 1
Rule1 5 => "once mode ON"

Rule2
  on rules#timer=1 do backlog var1 ON; publish sonnaz/%topic%/tele/PRESENCE_CHG ON; publish2 sonnaz/%topic%/tele/PRESENCE ON endon 
  on rules#timer=2 do backlog var1 OFF; publish sonnaz/%topic%/tele/PRESENCE_CHG OFF; publish2 sonnaz/%topic%/tele/PRESENCE OFF endon 
  on tele-VL53L1X do publish2 sonnaz/%topic%/tele/PRESENCE %var1% endon
Rule2 1
```

Rule1 is not trigger at teleperiod (there is no `tele-` prefix to the trigger) so it is executed in the sensor measurment loop. Which is likely every 1 or 2 seconds.
The timers are providing a debouncing (10 seconds) so that if someone cross under the sensor while there is no car in the garage, the change of presence will not be triggered (unless the person stand still right under the sensor which is on the ceiling)
Rule1 use the "once" mode enabled by `Rule1 5` so that it is not triggered unless the distance cross the threshold (%mem1%), currently 1m50.
I have 2 topics:
- PRESENCE which is a state periodically updated, and also updated on a detected change. It is using `Publish2` in order to be retained
- PRESENCE_CHG which is not retained and notify of a change of presence

You can simplify it without the debouncing timers if you don't care about notifying when some peoples on feet cross the sensor

I could have done the change detection in Node-RED too

This is very reactive as a change of distance is seen very quickly. Then the notification is delayed by the debouncing of course.

Sample of hourly teleperiod logs:
```
08:10:30.970 MQT: sonnaz/doorctr-garage1/tele/STATE = {"Time":"2024-02-09T08:10:30+01:00","Uptime":"14T15:03:33","UptimeSec":1263813,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":67,"POWER":"OFF","Wifi":{"AP":1,"SSId":"mado-wifi","BSSId":"80:AF:CA:05:41:80","Channel":10,"Mode":"11g","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:07"}}
08:10:31.103 MQT: sonnaz/doorctr-garage1/tele/SENSOR = {"Time":"2024-02-09T08:10:30+01:00","closed":"ON","opened":"OFF","not_used":"OFF","VL53L1X":{"Distance":212.4}} (retained)
08:10:31.120 RUL: TELE-VL53L1X performs "publish2 sonnaz/doorctr-garage1/tele/PRESENCE OFF"
08:10:31.214 MQT: sonnaz/doorctr-garage1/tele/PRESENCE = OFF (retained)
```

I'm not on site so I can't show you logs when the distance sensor is triggered.


Grafana dashboard from the influxDB data
1. door state (green is closed, orange is in between, red is open)
2. car presence
3. distance

![image](https://github.com/arendst/Tasmota/assets/2996042/aad1c15c-05aa-48b2-abb4-7d504faefbc0)


I hope this helps

Otherwise, please rpovde your rules and some logs when the sensor detects changes

**Reply by Noschvie** [reply]
`Rule1
  on rules#timer=1 do backlog var1 ON; publish sonnaz/%topic%/tele/PRESENCE_CHG ON; publish2 sonnaz/%topic%/tele/PRESENCE ON endon 
  on rules#timer=2 do backlog var1 OFF; publish sonnaz/%topic%/tele/PRESENCE_CHG OFF; publish2 sonnaz/%topic%/tele/PRESENCE OFF endon 
  on tele-VL53L1X do publish2 sonnaz/%topic%/tele/PRESENCE %var1% endon`

...should be Rule2 instead of Rule1, isn't it?

**Reply by barbudor** [reply]
Correct @Noschvie - I'm fixing it

**Comment by papakpmartin** [comment]
@barbudor, thank you for that detailed explanation of how you're doing it. I'm a little envious... I've not done anything with rules yet. I'm just working with the raw data in Home Assistant in a simple way, like this:

```yaml
binary_sensor:
- platform: template
  sensors:
    garage_door:
      friendly_name: Garage Door
      device_class: garage_door
      unique_id: garage_door
      value_template: >
        {{ is_number(states('sensor.garage_door_vl53l0x_distance')) 
           and states('sensor.garage_door_vl53l0x_distance') | int < 12 }} # inches
    car_in_garage:
      friendly_name: "Car in Garage"
      device_class: presence
      unique_id: car_in_garage
      value_template: >
        {% if is_number(states('sensor.garage_door_vl53l0x_distance')) %}
          {% if states('sensor.garage_door_vl53l0x_distance') | int < 60 %}
            {{ true }}
          {% else %}
            {{ false }}
          {% endif %}
        {% else %}
          {{ false }}
        {% endif %}
      availability_template: >
        {% if is_number(states('sensor.garage_door_vl53l0x_distance'))
              and states('sensor.garage_door_vl53l0x_distance') | int > 40 %}
          {{ true }}
        {% else %}
          {{ false }}
        {% endif %}
```

But in monkeying with this, I think I've narrowed down the problem. It is not anything to do with I2C. It appears to be that even though `garage_door_vl53l0x_distance` is updated very quickly at the device, the MQTT message doesn't seem to get fired off for a while. Leaving me in a state that looks like this, where reality is what the Tasmota is showing (about 6.5") and Home Assistant still only knows the distance from 30+ seconds ago:

<img width="1233" alt="Screen Shot 2024-02-09 at 3 23 54 PM" src="https://github.com/arendst/Tasmota/assets/3298248/e7ffa396-58ed-4f6d-91f3-e98359ab5bd9">

I don't think it's anything to do with my Mosquitto broker as other messages from other devices are flying around just fine. And even the "garage door button" message on this device is responsive.

So I2C was a distraction. Any thoughts (anyone) why some MQTT messages are responsive and some are not, on the same device?

**Reply by sfromis** [reply]
To see what happens, stop looking at HA, but instead the Tasmota console, where you can see all that's published via MQTT. And beware that by default, Tasmota is publishing sensor data by TelePeriod (default 5 minutes). As HA is not continuously getting updates, you should not expect values as current as what you see on the Tasmota front page.

**Reply by barbudor** [reply]
HA will not receive every measurements but only those that are published at teleperiod rate
Don't look at increasing this rate but instead learn how to detect changes on the device and published only when needed
Rules are your friends

**Reply by papakpmartin** [reply]
Thanks @sfromis and @barbudor. This all makes good sense. I think what threw me is that when I had this on the 8266 it did seem to be much more responsive and I had the same setup and the same lack of knowledge about rules.

Maybe earlier versions had a smaller default TelePeriod? Maybe I monkeyed with it and don't remember. Most likely the latter, I suppose.

Thanks for the help, all! Time to look at rules. :)

**Comment by papakpmartin** [comment, marked as answer]
Closing this up. The “answer” is the I2C is working fine and it's simply that I'm not getting the data sent via MQTT to Home Assistant as fast as I was before. Though I'm not _sure_ why, it's probably because I'd set something differently on the 8266 that I didn't on the ESP32.

**Reply by sfromis** [reply]
Well, `TelePeriod` allows intervals down to 10 seconds, which is not great if you want quick reaction time.
